### PR TITLE
feat: github client fetch support

### PR DESCRIPTION
**Pull request by lallevato-lm**
## Description

This change includes a client for Github with `fetch` functionality. For this client, `bomctl fetch` will retrieve the dependency graph sbom for the default commit of the specified repository. Currently Github does not support retrieving dependency graphs for other branches, commits, or releases.

Example usage:

`bomctl fetch https://github.com/bomctl/bomctl --netrc`

This command will fetch the dependency graph sbom for the bomctl repository.

Note: This change includes a `push.go` file for the Github client, but the file is basically an empty skeleton.

For context, see #132 

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: cmd/fetch.go [hunk 1]
```diff
@@ -37,8 +37,8 @@ func fetchCmd() *cobra.Command {
 	fetchCmd := &cobra.Command{
 		Use:   "fetch [flags] SBOM_URL...",
 		Args:  cobra.MinimumNArgs(fetchMinArgs),
-		Short: "Fetch SBOM file(s) from HTTP(S), OCI, or Git URLs",
-		Long:  "Fetch SBOM file(s) from HTTP(S), OCI, or Git URLs",
+		Short: "Fetch SBOM file(s) from HTTP(S), OCI, Git or Github URLs",
+		Long:  "Fetch SBOM file(s) from HTTP(S), OCI, Git, or Github URLs",
```
Comment (ashearin) [inline cmd/fetch.go:41]:
```suggestion
		Short: "Fetch SBOM file(s) from HTTP(S), OCI, Git, or Github URLs",
		Long:  "Fetch SBOM file(s) from HTTP(S), OCI, Git, or Github URLs",
```
Nit: Not sure if the intention was for them to be the same

File: internal/pkg/client/github/client_test.go [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+			name:     "http scheme with latest tag",
+			url:      "http://github.com/bomctl/bomctl",
+			owner:    "bomctl",
+			repoName: "bomctl",
+			expected: &netutil.URL{
+				Scheme:   "http",
+				Hostname: "github.com",
+				Path:     "bomctl/bomctl",
+			},
+		},
```
Comment (ashearin) [inline internal/pkg/client/github/client_test.go:88]: Should these have a tag value in the url?

File: internal/pkg/client/github/client_test.go [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+			url:      "git://git@github.com/bomctl/bomctl",
+			owner:    "bomctl",
+			repoName: "bomctl",
+			expected: &netutil.URL{
+				Scheme:   "git",
+				Username: "git",
+				Hostname: "github.com",
+				Path:     "bomctl/bomctl",
+			},
+		},
```
Comment (ashearin) [inline internal/pkg/client/github/client_test.go:239]: These two look like duplicates

File: internal/pkg/client/github/client_test.go (hunk 2 above)
Comment (lallevato-lm) [reply internal/pkg/client/github/client_test.go]: These were holdovers from the test file I copied it from. Removed.

File: internal/pkg/client/github/fetch.go [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+	ts := oauth2.StaticTokenSource(&oauth2.Token{AccessToken: auth.Password})
+	tc := oauth2.NewClient(ctx, ts)
+	client.ghClient = *github.NewClient(tc)
+
+	repoURL := strings.Split(url.Path, "/")
+	owner := repoURL[0]
+	repo := repoURL[1] + "/dependency-graph/sbom"
+	u := fmt.Sprintf("repos/%s/%s", owner, repo)
+
+	req, err := client.ghClient.NewRequest("GET", u, nil)
```
Comment (ashearin) [inline internal/pkg/client/github/fetch.go:56]: Would we not want to use the predefined sbom function here?

https://pkg.go.dev/github.com/google/go-github/v66/github#DependencyGraphService.GetSBOM

Comment (ashearin) [reply internal/pkg/client/github/fetch.go:56]: Looking at that GitLab command again, it returns a github sbom object, but you could use `.String()` to get what we nee I think

Comment (lallevato-lm) [reply internal/pkg/client/github/fetch.go:56]: For whatever reason, the `.String()` function returns a really weird format that protobom can't parse. The code in the MR just gets the raw JSON (as if you did a curl) and doesn't cast it to a github sbom objectm which just makes it a bit easier to work with.

**Review by ashearin — APPROVED** [review]

File: cmd/fetch.go [hunk 5]
```diff
@@ -37,8 +37,8 @@ func fetchCmd() *cobra.Command {
 	fetchCmd := &cobra.Command{
 		Use:   "fetch [flags] SBOM_URL...",
 		Args:  cobra.MinimumNArgs(fetchMinArgs),
-		Short: "Fetch SBOM file(s) from HTTP(S), OCI, or Git URLs",
-		Long:  "Fetch SBOM file(s) from HTTP(S), OCI, or Git URLs",
+		Short: "Fetch SBOM file(s) from HTTP(S), OCI, Git, or GitHub URLs",
+		Long:  "Fetch SBOM file(s) from HTTP(S), OCI, Git, or GitHub URLs",
```
[the remark this answers is not in the document]
Comment (lallevato-lm) [reply cmd/fetch.go:41]: Good call, I added an example.

File: internal/pkg/client/github/client.go [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+}
+
+func (*Client) Name() string {
+	return "github"
+}
+
+func (*Client) RegExp() *regexp.Regexp {
+	return regexp.MustCompile(
+		fmt.Sprintf("^%s%s%s%s$",
+			`(?P<scheme>https?|git|ssh):\/\/?`,
```
Comment (jhoward-lm) [inline internal/pkg/client/github/client.go:43]: The pattern `:\/\/?` would just mean an optional preceding forward slash. Did you mean this in order to make the whole scheme optional?

```suggestion
			`((?P<scheme>https?|git|ssh):\/\/)?`,
```

File: internal/pkg/client/github/client.go [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+			`((?P<username>[^:]+)(?::(?P<password>[^@]+))?(?:@))?`,
+			`(?P<hostname>github(\.[A-Za-z0-9_-]+)*\.com+)(?::(?P<port>\d+))?`,
+			`(?:[\/:](?P<path>[^@#]+)@?)`,
```
Comment (jhoward-lm) [inline internal/pkg/client/github/client.go:46]: If there is no segment of the URL following the path, then this pattern can probably be simplified to something like the following, or at the very least the `@?` at the end can be removed

```suggestion
			`(?:[\/:](?P<path>.+))`,
```

File: internal/pkg/client/github/fetch.go [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+	}
+
+	ts := oauth2.StaticTokenSource(&oauth2.Token{AccessToken: auth.Password})
+	tc := oauth2.NewClient(ctx, ts)
+	client.ghClient = *github.NewClient(tc)
+
+	repoURL := strings.Split(url.Path, "/")
+	owner := repoURL[0]
+	repo := repoURL[1] + "/dependency-graph/sbom"
+	u := fmt.Sprintf("repos/%s/%s", owner, repo)
```
Comment (jhoward-lm) [inline internal/pkg/client/github/fetch.go:54]: I'm surprised the linter is letting these slide. `ts`, `tc`, and `u` are too short for variable names. Can you rename them to `tokenSource`, `tokenClient`/`oauthClient`, and `apiPath`/`endpoint` or something like that?

File: internal/pkg/client/github/fetch.go [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+
+	req, err := client.ghClient.NewRequest("GET", u, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create request: %w", err)
+	}
+
+	req.Header.Add("Authorization", ("Bearer " + auth.Password))
```
Comment (jhoward-lm) [inline internal/pkg/client/github/fetch.go:61]:
```suggestion
	req.Header.Add("Authorization", "Bearer " + auth.Password)
```

**Review by jhoward-lm — CHANGES_REQUESTED** [review]

**Review by jhoward-lm — APPROVED** [review]

**Pull request merged**